File: game-engine/src/test/java/org/luxons/sevenwonders/game/resources/ResourceTransactionsTest.java
package org.luxons.sevenwonders.game.resources;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import org.junit.Test;
import org.luxons.sevenwonders.game.test.TestUtilsKt;

import static org.junit.Assert.assertEquals;

public class ResourceTransactionsTest {

    @Test
    public void toTransactions() {
        List<ResourceTransaction> transactionList = new ArrayList<>();
        transactionList.add(TestUtilsKt.createTransaction(Provider.LEFT_PLAYER, ResourceType.WOOD));
        transactionList.add(TestUtilsKt.createTransaction(Provider.LEFT_PLAYER, ResourceType.CLAY));
        transactionList.add(TestUtilsKt.createTransaction(Provider.RIGHT_PLAYER, ResourceType.WOOD));

        ResourceTransactions transactions = new ResourceTransactions(transactionList);

        Set<ResourceTransaction> expectedNormalized = new HashSet<>();
        expectedNormalized.add(
                TestUtilsKt.createTransaction(Provider.LEFT_PLAYER, ResourceType.WOOD, ResourceType.CLAY));
        expectedNormalized.add(TestUtilsKt.createTransaction(Provider.RIGHT_PLAYER, ResourceType.WOOD));

        assertEquals(expectedNormalized, new HashSet<>(transactions.toTransactions()));
    }

    @Test(expected = IllegalStateException.class)
    public void remove_failsIfNoResourceForThatProvider() {
        ResourceTransactions transactions = new ResourceTransactions();
        transactions.remove(Provider.LEFT_PLAYER, Resources.of(ResourceType.WOOD));
    }
}

File: game-engine/src/test/kotlin/org/luxons/sevenwonders/game/resources/ResourceTransactionsTest.kt
package org.luxons.sevenwonders.game.resources

import org.junit.Assert.assertEquals
import org.junit.Test
import org.luxons.sevenwonders.game.resources.ResourceType.CLAY
import org.luxons.sevenwonders.game.resources.ResourceType.WOOD
import org.luxons.sevenwonders.game.test.createTransaction

class ResourceTransactionsTest {

    @Test
    fun toTransactions() {
        val transactionMap = mapOf(
            Provider.LEFT_PLAYER to (1 of WOOD) + (1 of CLAY),
            Provider.RIGHT_PLAYER to (1 of WOOD)
        )

        val expectedNormalized = setOf(
            createTransaction(Provider.LEFT_PLAYER, WOOD, CLAY),
            createTransaction(Provider.RIGHT_PLAYER, WOOD)
        )

        assertEquals(expectedNormalized, transactionMap.toTransactions().toSet())
    }

    private infix fun Int.of(type: ResourceType): Resources = resourcesOf(type to this)
}
